### PR TITLE
Explore using configurers for composite properties

### PR DESCRIPTION
**Pull request by abstratt**
This is just a spike, not meant to be merged.

Issue: #18352 

RFC: https://docs.google.com/document/d/1KHCCzNeNAIkpedH9fUcNUe_vgZAWHxCODzxVW_7NzLw/edit#heading=h.610fausqnpu6

Proper PR: #27024 (WIP)

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/provider/HasMultipleValues.java [hunk 1]
```diff
@@ -77,52 +78,6 @@ public interface HasMultipleValues<T> extends HasConfigurableValue {
      */
     HasMultipleValues<T> empty();
 
-    /**
```
Comment (abstratt) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/HasMultipleValues.java:80]: These methods were pulled up into `CollectionConfigurer`.

File: subprojects/core-api/src/main/java/org/gradle/api/provider/MapProperty.java [hunk 2]
```diff
@@ -110,42 +111,6 @@ public interface MapProperty<K, V> extends Provider<Map<K, V>>, HasConfigurableV
      */
     MapProperty<K, V> value(Provider<? extends Map<? extends K, ? extends V>> provider);
 
-    /**
```
Comment (abstratt) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/MapProperty.java:113]: These were pulled up into `MapConfigurer`.

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: subprojects/core-api/src/main/java/org/gradle/api/provider/MapProperty.java [hunk 3]
```diff
@@ -188,4 +154,40 @@ public interface MapProperty<K, V> extends Provider<Map<K, V>>, HasConfigurableV
...
+    /**
+     * Updates the value of this property in place by executing the provided transformer.
+     * The transformer accepts the frozen value of this property.
+     *
+     * @param transformer the transformer to apply to frozen value of the property
+     * @return this
+     * @since 8.5
+     */
+    @Incubating
+    MapProperty<K, V> update(Transformer<? extends Provider<? extends Map<? extends K, ? extends V>>, ? super Provider<? extends Map<? extends K, ? extends V>>> transformer);
```
Comment (big-guy) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/MapProperty.java:167]: Does this need to use `Provider`s for the value or just the `Map` itself?  It seems like you can't do anything with a `Provider<Map<...>>` other than getting its value?

File: subprojects/core-api/src/main/java/org/gradle/api/provider/MapConfigurer.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     * Adds all entries from another {@link Map} to the property value.
+     *
+     * <p>The given provider will be queried when the value of this property is queried.
+     * This property will have no value when the given provider has no value.
+     *
+     * @param provider the provider of the entries
+     */
+    void putAll(Provider<? extends Map<? extends K, ? extends V>> provider);
+
+    void excludeAll(Predicate<K> keyFilter);
```
Comment (big-guy) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/MapConfigurer.java:70]: This should be a `Spec`

File: subprojects/core-api/src/main/java/org/gradle/api/provider/HasMultipleValues.java [hunk 5]
```diff
@@ -152,4 +133,15 @@ public interface HasMultipleValues<T> extends HasConfigurableValue {
      */
     @Override
     void finalizeValue();
+
+    /**
+     * Updates the value of this property in place by executing the provided transformer.
+     * The transformer accepts the frozen value of this property.
```
Comment (big-guy) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/HasMultipleValues.java:139]: Is "frozen" the same thing as finalized?

File: subprojects/core-api/src/main/java/org/gradle/api/provider/MapProperty.java (hunk 3 above)
Comment (abstratt) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/MapProperty.java:167]: Sorry for the noise, this change actually was brought in from @mlopatkin's https://github.com/gradle/gradle/pull/26634, please see his RFC on "Handling circular references in lazy configuration API".

File: subprojects/core-api/src/main/java/org/gradle/api/provider/HasMultipleValues.java (hunk 5 above)
Comment (abstratt) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/HasMultipleValues.java:139]: Also from @mlopatkin's PR, also explained here: https://github.com/gradle/gradle/pull/26564#discussion_r1375583508

File: subprojects/core-api/src/main/java/org/gradle/api/provider/MapConfigurer.java (hunk 4 above)
Comment (abstratt) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/MapConfigurer.java]: Done

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
Superceded by #21752 and others.

**Pull request closed without merging**